### ISSUE TITLE
Porting the work of this paper to a new dataset

### ISSUE THREAD

**Issue by yukaizhou** [astra-vision/MonoScene#39]
Hello author, first of all thank you for your great work. I want to directly apply your work to the nuscenes dataset, is it possible? Does the nuscenes dataset need point cloud data to assist in generating voxel data?

**Comment by anhquancao** [comment]
Hi @yukaizhou, Thank you! 
It's possible. In fact, I tried to infer on nuScenes using the model trained on Semantic KITTI, and it works fine, except some generalization issues. 
To generate the ground-truth voxels for nuScenes, you can follow the steps in this link:
https://github.com/yanx27/JS3C-Net#data-preparation
 
Here is my uncleaned code for loading the nuScenes dataset (inference only)

```
import torch
import os
import glob
from torch.utils.data import Dataset
import numpy as np
from nuscenes.nuscenes import NuScenes
from xmuda.data.utils.preprocess import create_img_grid, create_voxel_grid, select_points_in_frustum, vox2pixCamCoords, compute_local_frustums, compute_CP_mega_matrix, compute_local_frustums_enlarge
from PIL import Image
from torchvision import transforms

class NuscenesDataset(Dataset):
    def __init__(self, root):
        self.nusc = NuScenes(version='v1.0-mini', dataroot=root, verbose=True)
        self.img_H = 900
        self.img_W = 1600
        self.voxel_size = 0.2
        self.normalize_rgb = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
        ])

    def __len__(self):
        return len(self.nusc.sample)

    def __getitem__(self, index):
        data = {}
        sample = self.nusc.sample[index]
        sample_record = self.nusc.get('sample', sample['token'])
        camera_token = sample_record['data']['CAM_FRONT']
        cam_record = self.nusc.get('sample_data', camera_token)


        img = Image.open(os.path.join(self.nusc.dataroot, cam_record['filename']))
        data['img_path'] = cam_record['filename']
        img = np.array(img, dtype=np.float32, copy=False) / 255.
        img = self.normalize_rgb(img)
        data['img'] = img

        calib_record = self.nusc.get('calibrated_sensor', cam_record['calibrated_sensor_token'])
        cam_k = np.array(calib_record['camera_intrinsic'])
        T_velo_2_cam = np.identity(4) # reconstruct in camera coordinates 

        data['T_velo_2_cam'] = T_velo_2_cam
        data['cam_k'] = cam_k
        data['frame_id'] =  str(index)

        scales = [1, 2]
        data['scales'] = scales
        vox_origin = np.array([-25.6, -4, 0])
        for scale in scales:
            pix, valid_pix = vox2pixCamCoords(cam_k, self.voxel_size, scale, self.img_W, self.img_H)
            data['pix_' + str(scale)] = pix
            data['valid_pix_' + str(scale)] = valid_pix
        return data


if __name__ == '__main__':
    nuscenes_ds = NuscenesDataset('/gpfsscratch/rech/kvd/uyl37fq/dataset/nuscenes')
    for i in range(len(nuscenes_ds)):
        print(list(nuscenes_ds[i].keys()))
```

**Comment by yukaizhou** [comment]
Thank you very much for your prompt reply, I will try your suggestion. Thanks again

**Comment by yukaizhou** [comment]
Hello, according to the repo you provided（https://github.com/jbehley/voxelizer/tree/mast）, when generating the voxels of nuscenes, it has been unable to configure the environment successfully. Do you have any other methods? Thank you for your guidance.

**Comment by anhquancao** [comment]
Hi, sorry, I have no other methods in mind.

**Comment by yukaizhou** [comment]
Hello, I would like to ask you a question. When I use voxel to organize, I use lidar-seg as the voxel label information. What should I use as the origin of the coordinate system at this time? Ego or pixel coordinate system? In addition, I have tried to convert the lidar-seg to the pixel coordinate system, but the image size of the nuscenes is 900*1600. At this time, when the voxel interval is 2mm, the created voxel volume will be very large? Another way is to use the self-vehicle as the origin of the coordinate system, but at this time, when it is point cloud coordinates (x, y, z, vis), it is found that z, that is, the depth, is less than 0, so what I want to confirm is that it is less than 0 Is it the point cloud in the field of view of the 3 cameras 'BACK', 'BACK_LEFT', 'BACK_RIGHT'? Looking forward to thank you for your answer. thanks.

**Comment by anhquancao** [comment]
https://www.cvlibs.net/datasets/kitti/setup.php

**Comment by anhquancao** [comment]
Hi,
For the detail of the data generation you should check this paper
https://arxiv.org/abs/1904.01416
To generate the ground-truth, we don't need the image so the image is irrelevnt. We only need the lidar points. The voxelized scene is the part of the points cloud in the area with x = [0, 51.2], y=[-25.6, 25.6], z=[-2, 4.4] in front of the vehicle so z > 0.
The origin of the lidar voxel in kitti is [x, y, z] = [0. -25.6, -2]

**Comment by yukaizhou** [comment]
Hello, I have obtained the voxel volume by using the point cloud data of nuscenes according to your guidance, where (self.voxel_size = 0.2; self.scene_size = (51.2, 51.2, 6.4); self.vox_origin = np.array([0 , -25.6, -2])). But there are still some questions, please guide me. 1. What is the function of compute_CP_mega_matrix? How should I modify the nuscenes? Thanks again for your guidance. thanks

**Comment by anhquancao** [comment]
Do you mean this function? 
https://github.com/astra-vision/MonoScene/blob/22256acbf452f38395c28c180102bc6942699a2d/monoscene/data/utils/helpers.py#L6
It is used to compute the voxel-voxel relation, as described in Sec. 3.2 in the main paper.
